lib/ui/painting/canvas.cc

```diff
@@ -94,18 +94,23 @@ Canvas::Canvas(SkCanvas* canvas) : canvas_(canvas) {}
 Canvas::~Canvas() {}
 
 void Canvas::save() {
-  if (!canvas_) {
-    return;
+  if (display_list_recorder_) {
+    builder()->save();
```

Comment:
I find the way `Canvas` vs. `DisplayListCanvasRecorder` is distinguished here to be pretty confusing. In the case of the `DisplayListCanvasRecorder` being enabled, we end up with both `canvas_` and `display_list_recorder_` being the same object, as I see it?

It seems to me that if `DisplayListCanvasRecorder` is actually implementing the `SkCanvas` API then there should be no reason to special case each of the methods in `Canvas` that operate on either an `SkCanvas` or a `DisplayListCanvasRecorder`. It seems more logical to me that during construction we'd either set `canvas_` to the `SkCanvas` object or the `DisplayListCanvasRecorder` object, then they handle the semantic differences in their own implementations of the `SkCanvas` API methods that are called by `Canvas`, instead of `Canvas` needing to special case depending on which class it actually is.

Reply:
Ah, I should fix the construction here. We had been using the canvas_ for all rendering and so the construction code always set it up, but now that we are not using it for DL, I should have conditional construction that sets up one or the other, but not both.

Reply:
Can't get rid of it for now. The canvas is used by the paragraph code to inject text. It's not clear how much work will be involved in getting rid of that. :(

See: https://github.com/flutter/engine/blob/9505dc4e03a0c8defbb75f14fe5822c57c5d6cd4/lib/ui/text/paragraph.cc#L95

Reply:
I added a comment to the initialization code to document this implicit dependency.